### PR TITLE
[Bugfix] V1: fix allowed_token_ids_mask aliasing in InputBatch.swap_states

### PR DESCRIPTION
**Pull request by huthvincent**
## Purpose

`InputBatch.swap_states()` swaps the two `allowed_token_ids_mask_cpu_tensor`
rows with a tuple assignment on row **views**:

```python
(
    self.allowed_token_ids_mask_cpu_tensor[i1],
    self.allowed_token_ids_mask_cpu_tensor[i2],
) = (
    self.allowed_token_ids_mask_cpu_tensor[i2],
    self.allowed_token_ids_mask_cpu_tensor[i1],
)
```

Integer indexing on a `torch.Tensor` returns a **view**, not a copy. The RHS
tuple holds two views into the live tensor, so the left-to-right assignment
first overwrites row `i1` with row `i2`, then overwrites row `i2` with the
row-`i1` view that now already contains row `i2`'s data. Both rows collapse to
the original `i2` contents, and the request that ends up at index `i2` silently
loses its `allowed_token_ids` restriction (or inherits the wrong one).

`swap_states` is invoked by `reorder_batch`
(`vllm/v1/attention/backends/utils.py`) on the MLA / Mamba / GDN / FlashInfer
decode–prefill split, so any request using `SamplingParams.allowed_token_ids`
that gets reordered samples from the wrong token set — wrong output, no error.

The fix uses fancy indexing (which copies on the RHS), matching the safe
`is_token_ids` swap two lines above. Only the CPU staging tensor needs swapping;
`_make_sampling_metadata` copies it to the GPU tensor every step.

**Not a duplicate.** Issue #43894 / PR #43931 address the sibling row-leak in
`InputBatch.condense()`. This PR fixes the separate `swap_states` aliasing that
none of them touch.

## Test Plan

Added `test_swap_states_preserves_allowed_token_ids_mask` to
`tests/v1/worker/test_gpu_input_batch.py` (constrained request at row 0,
unconstrained at row 1, swap, assert the constraint follows the request).

```bash
pytest tests/v1/worker/test_gpu_input_batch.py::test_swap_states_preserves_allowed_token_ids_mask
pytest tests/v1/worker/test_gpu_input_batch.py
```

## Test Result

Verified on 1× NVIDIA H200 (CUDA 13.0, torch 2.11.0).

```
# new test, BEFORE fix (bug reproduced):
>   assert int(mask[1].sum().item()) == VOCAB_SIZE - 1
E   assert 0 == (1024 - 1)
1 failed

# new test, AFTER fix:
1 passed

# whole file, AFTER fix:
10 passed

# lint:
ruff check      -> All checks passed!
ruff format     -> 2 files already formatted
```

The change only alters behavior for `allowed_token_ids` requests that are
reordered; it restores the intended masking, so it can only make constrained
output more correct. Covered by the deterministic unit test above.

---

*AI assistance (Claude) was used to develop this change. The submitter has
reviewed every changed line and run the tests above.*

### CODE REVIEW

**Comment by ErenAta16** [comment]
Verified the root cause independently -- confirmed the aliasing with a quick numpy repro (1D `arr[i1], arr[i2] = arr[i2], arr[i1]` is safe since scalar indexing copies, but the 2D row-view case collapses both rows to the second one's data exactly as described):

```python
>>> import numpy as np
>>> a = np.arange(12.).reshape(3, 4)
>>> a[0], a[1] = a[1], a[0]
>>> a
array([[ 4.,  5.,  6.,  7.],
       [ 4.,  5.,  6.,  7.],   # row 0 leaked into row 1
       [ 8.,  9., 10., 11.]])
```

Given `swap_states` has ~15 other tuple-swapped fields using the identical `self.X[i1], self.X[i2] = self.X[i2], self.X[i1]` pattern, I went through all of them to check whether any others share this bug:

- `num_tokens_no_spec`, `num_prompt_tokens`, `num_computed_tokens_cpu`, `temperature_cpu`, `top_p_cpu`, `top_k_cpu`, `frequency_penalties_cpu`, `presence_penalties_cpu`, `repetition_penalties_cpu`, `num_accepted_tokens_cpu`, `request_lora_mapping` are all **1D** `(max_num_reqs,)` arrays -- basic integer indexing on a 1D array returns a scalar copy, not a view, so these are safe.
- `req_output_token_ids`, `spec_token_ids` are plain Python lists; `req_id_to_index` is a dict -- also unaffected (no view semantics).
- `token_ids_cpu` (2D) and `is_token_ids` (2D) already use the safe copy/fancy-index pattern, per the existing `NOTE: the following is unsafe` comment.

So `allowed_token_ids_mask_cpu_tensor` (2D: `(max_num_reqs, vocab_size)`) is the only remaining field with this bug class in `swap_states` -- the fix is complete, not partial. Good catch, and the fancy-indexing fix matches the existing `is_token_ids` pattern exactly, so no new failure mode introduced. Test looks right too: it pins one row to a single allowed token and the other unconstrained, which is exactly the case that collapses under the old tuple-of-views assignment.

**Comment by huthvincent** [comment]
Thanks for the thorough independent verification, @ErenAta16 — especially auditing the other ~15 swapped fields to confirm the 2D mask is the only affected one. That matches what I found: the 1D `_cpu` arrays return scalar copies and `token_ids_cpu`/`is_token_ids` already use the safe fancy-index pattern, so this completes that bug class in `swap_states`.

**Comment by ErenAta16** [comment]
Glad it lines up. Good find on your end too -- this is a genuinely easy pattern to miss since it only bites on 2D fields, and most of the file's swapped state is 1D.

**Comment by huthvincent** [comment]
@njhill when you have a moment — this is a small, self-contained V1 fix (`allowed_token_ids_mask` view-aliasing in `InputBatch.swap_states`) with a regression test, plus independent verification from @ErenAta16 that the 2D mask is the only affected field. As a first-time contributor I can't trigger CI myself — could you add the `ready` label so it can run? Thanks!

**Comment by huthvincent** [comment]
Hi @njhill, a gentle follow-up on this one. I re-checked against current main (80389cf): the tuple-of-views swap is still at gpu_input_batch.py:692-699, and the patch still applies cleanly after #48018 and #53464, so no rebase is needed. I understand MRv1 is low priority now that #53183 makes MRv2 the default; since MRv1 is still selected for the CPU backend (no-Triton installs), the ROCm DSA models, ngram spec decode and custom logits processors, I thought this small fix plus regression test might still be worth landing. If you would rather not take MRv1-only fixes at this point, I am happy to close it. Thanks either way!

**Comment by hyunnnchoi** [comment]
Hey, I independently verified this.  the bug still reproduces on latest main.

As of `5efbf0886a`, the tuple-of-views swap is still there at `gpu_input_batch.py:689-696`. I ran the actual swap_states() source on both CPU and CUDA (A30), and both rows end up with row i2's contents:

```
before row0 [1,1,1,0,0,0,0,0] row1 [0,0,0,1,1,1,1,1]
after  row0 [0,0,0,1,1,1,1,1] row1 [0,0,0,1,1,1,1,1]
```

What makes this worse is that `reorder_batch_to_split_decodes_and_prefills` calls `swap_states` repeatedly while following swap cycles, so a single reorder step can spread the corruption across multiple rows. The affected requests then silently sample from the wrong allowed-token set — no error, just wrong results.

The existing `test_swap_states_in_input_batch` doesn't catch this because `_create_sampling_params()` never sets `allowed_token_ids`, so the mask tensor is never allocated. 

This is MRv1-only, but it's still a silent correctness issue for any configuration that falls back to the V1 runner (e.g. spec decode methods MRv2 doesn't support yet, or certain ROCm architectures). Happy to also add test cases covering two requests with different allowed-token sets or the reorder path if that'd help.

**Review by njhill — APPROVED** [review]
Thanks @huthvincent
